teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,420 @@
+
+Each `artifact` is described by two fields, `uri` and `digest`. The `uri` includes the artifact `name`, as shown in the [`uri specification`](#uri). An overview of how `uri` and `digest` are used by steps is provided below:
+
+- `download` step: [(`uri`, expected `digest`)] -> [`/tekton/artifacts/downloads/<digest>.tgz`]
+- `verify` step: [`/tekton/artifacts/downloads/<digest>.tgz`] -> [(`verified`, `/tekton/artifacts/inputs/<name>/`)]
+- `digest` step: [`<name>`] -> [(`uri`, `digest`, `/tekton/artifacts/uploads/<name>.tgz`)]
+- `upload` step: [`/tekton/artifacts/outputs/<name>.tgz`] -> [(`uri`, `digest`)]
+
+The full details are described in the [step implementation section](#steps-implementation).
+
+The data in `uri` and `digest` could be used by steps to produce provenance data for input artifacts (aka "materials") and output artifacts (aka "products") to be added to `TaskRun.status.provenance`. This functionality is not the scope for this TEP, but the design takes it into account.
+
+#### `uri`
+
+The `uri` component identifies an artifact within the artifact storage in use.
+The storage itself is included as part of the `TaskRun` and `PipelineRun` definition.
+
+The `uri` is in [purl-generic](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic). Example values:
+
+- `pkg:generic/<name>`
+- `pkg:generic/build`
+- `pkg:generic/repo`
+- `pkg:generic/test-results`
+- `pkg:generic/sbom`
+
+
+#### `digest`
+
+The `digest` component includes the result of applying an hash function to the artifact itself.
+The `digest` is in the format `hash-function:<digest>`. Example values:
+
+- `md5:471a3a9ddc1d6e699190b9f4ca88aadf`
+- `sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2`
+
+### Steps Implementation
+
+Steps in Tekton do not have a YAML-based API. This may change one [TEP-0142 - Reusable Steps](https://github.com/tektoncd/community/pull/1065) is implemented, but only for reusable steps. Inlined steps will continue same as today.
+
+Even without a YAML API, the step interface can be defined as a combination of input environment variables and files and output files.
+Tekton will then map data coming from the Tekton API into the `download` step interface and data produced by the `upload` and `download` steps into the `TaskRun` status.
+
+This approach enables users who do not wish to use the [artifacts API](#artifacts-api) to use custom steps and still have provenance data injected into the `TaskRun` status.
+
+Steps are implemented as a single go binary that takes command line arguments to implement the logic of the four different steps.
+The go binary is built for all Tekton supported architectures and OSes. The steps point to the multi-arch image manifest so that the container runtime may download the correct version.
+
+#### Digest
```

Comment:
How is the metadata(digest and uri) information surfaced to the TaskRun Status? 
- Termination Message?
- Sidecar logs?
- might need both.

Reply:
It's on L253 

> The artifact provenance data is extracted from the local storage using the same mechanism used for results. Results and artifacts data is collated together and stored either in the termination message or in a sidecar log, depending on configuration.

Reply:
Thanks! I probably missed it. Feel free to resolve!